cql3/restrictions/single_column_primary_key_restrictions.hh

```diff
@@ -109,14 +113,16 @@ public:
         , _allow_filtering(other._allow_filtering)
         , _restrictions(::make_shared<single_column_restrictions>(schema))
     {
-        for (const auto& entry : other._restrictions->restrictions()) {
+        for (const auto& entry : other.restrictions()) {
             const column_definition* other_cdef = entry.first;
             const column_definition* this_cdef = _schema->get_column_definition(other_cdef->name());
             if (!this_cdef) {
                 throw exceptions::invalid_request_exception(format("Base column {} not found in view index schema", other_cdef->name_as_text()));
             }
-            ::shared_ptr<single_column_restriction> restriction = entry.second;
-            _restrictions->add_restriction(restriction->apply_to(*this_cdef));
+            // Make a(ny) concrete subclass, since single_column_restriction is still abstract for now.
+            auto r = ::make_shared<single_column_restriction::EQ>(*this_cdef, nullptr);
+            r->expression = replace_column_def(entry.second->expression, this_cdef);
```

Comment:
So we created an expression without a column definition, then substitute it after the fact?

Is there no way to avoid this very surprising behavior?

Reply:
> So we created an expression without a column definition,

Not without, but with the original column.

> then substitute it after the fact?
> 
> Is there no way to avoid this very surprising behavior?

Instead of replace_column_def, we could construct a morphologically identical expression (but with substituted column_def everywhere) from scratch.  That would be more code that needs updating whenever the expression hierarchy changes.

Reply:
I don't understand how we got to an expression that had the wrong column_definition in the first place. Once we are in this situation, I can understand using replace. But I don't understand how we got here.

Reply:
I think I see now - schema conversion. Let me see how why we need it in the first place.

Reply:
This is used by MV...

Reply:
And the whole thing happens at query time instead of prepare time!

Reply:
So the way it should work is that an select statement's prepare should construct a new expression/restriction during prepare time. But we're not going to change that now, so let's leave it as is. But wow, is this thing super tangled or what.

Reply:
> But wow, is this thing super tangled or what

Hyper-tangled.  Welcome to my life. :smile: